ambari-server/src/main/java/org/apache/ambari/server/configuration/AmbariServerConfigurationProvider.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.ambari.server.configuration;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.apache.ambari.server.events.AmbariConfigurationChangedEvent;
+import org.apache.ambari.server.events.JpaInitializedEvent;
+import org.apache.ambari.server.events.publishers.AmbariEventPublisher;
+import org.apache.ambari.server.orm.GuiceJpaInitializer;
+import org.apache.ambari.server.orm.dao.AmbariConfigurationDAO;
+import org.apache.ambari.server.orm.entities.AmbariConfigurationEntity;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.eventbus.Subscribe;
+import com.google.inject.Inject;
+import com.google.inject.Provider;
+
+/**
+ * AmbariServerConfigurationProvider is an abstract class to be extended by Ambari server configuration
+ * provider classes.
+ * <p>
+ * This implementation registers for the following Ambari server events to help load the relevant properties
+ * when available and/or updated:
+ *
+ * <ul>
+ * <li>AmbariConfigurationChangedEvent</li>
+ * <li>JpaInitializedEvent</li>
+ * </ul>
+ */
+public abstract class AmbariServerConfigurationProvider<T extends AmbariServerConfiguration> implements Provider<T> {
+  private static final Logger LOGGER = LoggerFactory.getLogger(AmbariServerConfigurationProvider.class);
+
+  @Inject
+  private Provider<AmbariConfigurationDAO> ambariConfigurationDAOProvider;
+
+  private final AtomicBoolean jpaInitialized = new AtomicBoolean(false);
+
+  private final AmbariServerConfigurationCategory configurationCategory;
+
+  private T instance = null;
+
+  protected AmbariServerConfigurationProvider(AmbariServerConfigurationCategory configurationCategory, AmbariEventPublisher publisher, GuiceJpaInitializer guiceJpaInitializer) {
```

Comment:
Any reason you chose to have the implementations pass in the injected instance instead of injecting it directly?

Reply:
I was having issues with the unit tests and this seemed to be the path that wound up working the best.  Now that I was able to solve the unit test issues, I can probably change it to be injected directly. However it seemed like the injection may not have been complete by the time the constructor was called and therefore using the injected members in the constructor was risky. This is especially true if partially mocking an `AmbariServerConfgurationProvider` in a unit test.

Reply:
OK, we can leave it if it was giving you UT problems. It's cleaner to have it injected so each implementation doesn't need to worry about it. What about injecting a `Provider<>` of it instead? That usually helps fix circular dependency and timing problems.

Reply:
I tried moving the `AmbariEventPublisher` and `GuiceJpaInitializer` parameters to be injected provider members
```
  @Inject
  private Provider<AmbariEventPublisher> ambariEventPublisherProvider;

  @Inject
  private Provider<GuiceJpaInitializer> guiceJpaInitializerProvider;
```

However, they were `null` when the `AmbariAuthenticationProvider` constructor was executed.  This caused an NPE.

Reply:
Is the provided `@Inject` annotated?

Reply:
`AmbariServerConfigurationProvider`'s constructor is not `@Inject` annotated, but the derived classes' constructors are:

```
  @Inject
  public AmbariLdapConfigurationProvider(AmbariEventPublisher publisher, GuiceJpaInitializer guiceJpaInitializer) {
```

Reply:
OK, not an issue we need to dwell on :)

Reply:
Awesome.. thanks! :)